server/mailrepository/mailrepository-api/src/test/java/org/apache/james/mailrepository/MailRepositoryContract.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    default  Mail createMail(String name, String body) throws MessagingException {
+        MimeMessage mailContent = generateMailContent(body);
+        List<MailAddress> recipients = ImmutableList
+            .of(new MailAddress("rec1@domain.com"),
+                new MailAddress("rec2@domain.com"));
+        MailAddress sender = new MailAddress("sender@domain.com");
+        Mail mail = new MailImpl(name, sender, recipients, mailContent);
+        mail.setAttribute(TEST_ATTRIBUTE, "testValue");
+        return mail;
+    }
+
+
+    default MimeMessage generateMailContent(String body) throws MessagingException {
+        return MimeMessageBuilder.mimeMessageBuilder()
+            .setSubject("test")
+            .setText(body)
+            .build();
+    }
+
+    default void checkMailEquality(Mail actual, Mail expected) {
+        assertAll(
+            () -> assertThat(actual.getMessage().getContent()).isEqualTo(expected.getMessage().getContent()),
+            () -> assertThat(actual.getMessageSize()).isEqualTo(expected.getMessageSize()),
+            () -> assertThat(actual.getName()).isEqualTo(expected.getName()),
+            () -> assertThat(actual.getState()).isEqualTo(expected.getState()),
+            () -> assertThat(actual.getAttribute(TEST_ATTRIBUTE)).isEqualTo(expected.getAttribute(TEST_ATTRIBUTE))
+        );
+    }
+
+    MailRepository retrieveRepository() throws Exception;
+
+    @Test
+    default void storeRegularMailShouldNotFail() throws Exception {
+        MailRepository testee = retrieveRepository();
+        Mail email = createMail("mail1");
+
+        testee.store(email);
+    }
+
+    @Test
+    default void retrieveShouldGetStoredMail() throws Exception {
+        MailRepository testee = retrieveRepository();
+        String key1 = "mail1";
+        Mail email = createMail(key1);
+
+        testee.store(email);
+
+        assertThat(testee.retrieve(key1)).satisfies(actual -> checkMailEquality(actual, email));
+    }
+
+    @Test
+    default void newlyCreatedRepositoryShouldNotContainAnyMail() throws Exception {
+        MailRepository testee = retrieveRepository();
+
+        assertThat(testee.list()).isEmpty();
+    }
+
+    @Test
+    default void retrievingUnknownMailShouldReturnNull() throws Exception {
+        MailRepository testee = retrieveRepository();
+
+        assertThat(testee.retrieve("random")).isNull();
+    }
+
+    @Test
+    default void removingUnknownMailShouldHaveNoEffect() throws Exception {
+        MailRepository testee = retrieveRepository();
+
+        testee.remove("random");
+    }
+
+    @Test
+    default void listShouldReturnStoredMailsKeys() throws Exception {
+        MailRepository testee = retrieveRepository();
+        String key1 = "mail1";
+        String key2 = "mail2";
+        testee.store(createMail(key1));
+
+        testee.store(createMail(key2));
+
+        assertThat(testee.list()).containsExactly(key1, key2);
```

Comment:
Isn't exactly a bit tough?

(How do you define order?)

Please relax that assertion.